File: golang/internal/server/rest/regex_safe.go
package rest

import "strings"

const maxRegexPatternLen = 256

// SafeRegexPattern rejects oversized or obviously catastrophic patterns before compile.
func SafeRegexPattern(pattern string) (string, bool) {
	if pattern == "" || len(pattern) > maxRegexPatternLen {
		return "", false
	}
	// ponytail: heuristic only; Go regexp is RE2 (linear), but cap pattern size for parity with TS.
	if strings.Contains(pattern, "(?") && strings.ContainsAny(pattern, "+*{") {
		return "", false
	}
	if nestedQuantifierPattern(pattern) {
		return "", false
	}
	return pattern, true
}

func nestedQuantifierPattern(pattern string) bool {
	for i := 0; i < len(pattern)-2; i++ {
		if pattern[i] != '(' {
			continue
		}
		depth := 1
		for j := i + 1; j < len(pattern); j++ {
			switch pattern[j] {
			case '(':
				depth++
			case ')':
				depth--
				if depth == 0 {
					inner := pattern[i+1 : j]
					if strings.ContainsAny(inner, "+*") && j+1 < len(pattern) && strings.ContainsAny(pattern[j+1:], "+*{") {
						return true
					}
					break
				}
			}
		}
	}
	return false
}

File: src/lib/wow/formats/regex-safe.ts
const MAX_REGEX_PATTERN_LEN = 256;

export function safeRegexPattern(pattern: string): string | null {
  if (!pattern || pattern.length > MAX_REGEX_PATTERN_LEN) return null;
  if (nestedQuantifierPattern(pattern)) return null;
  try {
    // Compile check only — actual search uses the validated pattern.
    // eslint-disable-next-line no-new
    new RegExp(pattern, 'i');
    return pattern;
  } catch {
    return null;
  }
}

function nestedQuantifierPattern(pattern: string): boolean {
  for (let i = 0; i < pattern.length - 2; i++) {
    if (pattern[i] !== '(') continue;
    let depth = 1;
    for (let j = i + 1; j < pattern.length; j++) {
      if (pattern[j] === '(') depth++;
      else if (pattern[j] === ')') {
        depth--;
        if (depth === 0) {
          const inner = pattern.slice(i + 1, j);
          if (/[+*/]/.test(inner) && j + 1 < pattern.length && /[+*}]/.test(pattern.slice(j + 1))) {
            return true;
          }
          break;
        }
      }
    }
  }
  return false;
}
